plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,495 @@
+
+-- | Shriking-order on kinds.
+--
+-- A kind `k1 = foldr (->) * ks1` is less than or equal to a kind `k2 = foldr (->) * ks2`
+-- when `ks1` can be constructed by dropping and shrinking kinds in `ks2`.
+--
+-- This shrinking order means that `* -> (* -> * -> * -> *) -> *` can shrink to
+-- `* -> *` or `* -> (* -> *) -> *` but not to `* -> * -> * -> *`. Not allowing
+-- this final shrink is important as we are guaranteed to only ever reduce the
+-- number of type arguments we need to provide when shrinking - thus allowing us
+-- to guarantee that e.g. generated datatypes never increase their number of
+-- parameters. This restriction is important because once the number of parameters
+-- starts to grow during shrinking it is difficult to ensure that the size of
+-- generated types and terms doesn't baloon and cause a shrink-loop.
+leKind :: Kind () -> Kind () -> Bool
+leKind k1 k2 = go (reverse $ argsKind k1) (reverse $ argsKind k2)
+  where
+    go [] _                = True
+    go _ []                = False
+    go (k : ks) (k' : ks')
+      | leKind k k' = go ks ks'
+      | otherwise   = go (k : ks) ks'
+
+-- | Strict shrinking order on kinds
+ltKind :: Kind () -> Kind () -> Bool
+ltKind k k' = k /= k' && leKind k k'
+
+-- | Take a type in a context and a new target kind
+--   Precondition: new kind is smaller or equal to old kind of the type.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => Map TyName (Kind ())
+        -> Type TyName DefaultUni ()
+        -> Kind ()
+        -> Type TyName DefaultUni ()
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | unsafeInferKind ctx ty == k = ty
+  | not $ k `leKind` unsafeInferKind ctx ty =
+      error "fixKind not smaller"
+  | otherwise = case ty of
+    -- Switch a variable out for a different variable of the right kind
+    TyVar _ _ | y : _ <- [ y | (y, k') <- Map.toList ctx
+                             , k == k' ] -> TyVar () y
+              | otherwise -> minimalType k
+    -- Try to fix application by fixing the function
+    TyApp _ a b       -> TyApp () (fixKind ctx a $ KindArrow () (unsafeInferKind ctx b) k) b
+    TyLam _ x kx b    ->
+      case k of
+        -- Fix lambdas to * by substituting a minimal type for the argument
+        -- and fixing the body.
+        Type{}        -> fixKind ctx (substClosedType x (minimalType kx) b) k
+        -- Fix functions by either keeping the argument around (if we can) or getting
+        -- rid of the argument (by turning its use-sites into minimal types) and introducing
+        -- a new argument.
+        KindArrow _ ka kb
+          | ka == kx              -> TyLam () x kx $ fixKind (Map.insert x kx ctx) b kb
+          | not $ kb `leKind` kb' -> error "fixKind"
+          | otherwise             -> TyLam () x ka $ fixKind ctx' b' kb
+            where
+              ctx' = Map.insert x ka ctx
+              b'   = substClosedType x (minimalType kx) b
```

Comment:
I can't see a reason why not off the top of my head - but it's worth while checking for shrink-loops very carefully if you implement that.

Reply:
Well the thing is that since the shrinking order on types isn't specified, I can't analytically check for shrinking loops and empirical evidence isn't particularly reliable.

Reply:
I'm not sure if I agree that empirical evidence isn't reliable... You can always run a few million tests and be pretty confident!

Reply:
OK, so I've played with it and it didn't work in the end. I left the following comment:

```
        -- We could've used @fixKind (Map.insert x ka' ctx) (TyVar () x) ka)@ here instead of
        -- @minimalType ka@, so that the usages of @x@ are preserved when possible, but that would
        -- mean fixing a kind to a bigger one (because @ka'@ has to be smaller than @ka@ and we'd go
        -- in the opposite direction), which is not supported by 'fixKind' (even though just
        -- commenting out the relevant check and making the change here does seem to give us
        -- better shrinking that still properly terminates, 'cause why would it not).
```

I think it's something worth improving in future, but not gonna make it a blocker for the time being.